src/plots/gl3d/scene.js

```diff
@@ -59,40 +59,67 @@ function render(scene) {
     }
 
     function formatter(axisName, val) {
-        if(val === undefined) return undefined;
         if(typeof val === 'string') return val;
 
         var axis = scene.fullSceneLayout[axisName];
         return Axes.tickText(axis, axis.c2l(val), 'hover').text;
     }
 
+    function makeEventData(xVal, yVal, zVal, trace, selection) {
+        return {points: [{
+            x: xVal,
+            y: yVal,
+            z: zVal,
+            data: trace._input,
+            fullData: trace,
+            curveNumber: trace.index,
+            pointNumber: selection.data.index
+        }]};
+    }
+
+    var oldEventData;
+
     if(lastPicked !== null) {
         var pdata = project(scene.glplot.cameraParams, selection.dataCoordinate),
-            hoverinfo = lastPicked.data.hoverinfo;
+            trace = lastPicked.data,
+            hoverinfo = trace.hoverinfo;
+
+        var xVal = formatter('xaxis', selection.traceCoordinate[0]),
+            yVal = formatter('yaxis', selection.traceCoordinate[1]),
+            zVal = formatter('zaxis', selection.traceCoordinate[2]);
 
         if(hoverinfo !== 'all') {
             var hoverinfoParts = hoverinfo.split('+');
-            if(hoverinfoParts.indexOf('x') === -1) selection.traceCoordinate[0] = undefined;
-            if(hoverinfoParts.indexOf('y') === -1) selection.traceCoordinate[1] = undefined;
-            if(hoverinfoParts.indexOf('z') === -1) selection.traceCoordinate[2] = undefined;
+            if(hoverinfoParts.indexOf('x') === -1) xVal = undefined;
+            if(hoverinfoParts.indexOf('y') === -1) yVal = undefined;
+            if(hoverinfoParts.indexOf('z') === -1) zVal = undefined;
             if(hoverinfoParts.indexOf('text') === -1) selection.textLabel = undefined;
             if(hoverinfoParts.indexOf('name') === -1) lastPicked.name = undefined;
         }
 
         Fx.loneHover({
-            x: (0.5 + 0.5 * pdata[0]/pdata[3]) * width,
-            y: (0.5 - 0.5 * pdata[1]/pdata[3]) * height,
-            xLabel: formatter('xaxis', selection.traceCoordinate[0]),
-            yLabel: formatter('yaxis', selection.traceCoordinate[1]),
-            zLabel: formatter('zaxis', selection.traceCoordinate[2]),
+            x: (0.5 + 0.5 * pdata[0] / pdata[3]) * width,
+            y: (0.5 - 0.5 * pdata[1] / pdata[3]) * height,
+            xLabel: xVal,
+            yLabel: yVal,
+            zLabel: zVal,
             text: selection.textLabel,
             name: lastPicked.name,
             color: lastPicked.color
         }, {
             container: svgContainer
         });
+
+        var eventType = (selection.buttons) ? 'plotly_click' : 'plotly_hover',
+            eventData = makeEventData(xVal, yVal, zVal, trace, selection);
+
+        scene.graphDiv.emit(eventType, eventData);
```

Comment:
To note, click events are fired within the same distance than hover events. Specifically, click event can be fired in situation where the cursor isn't exactly on top of a given data point.

This behaviour is debatable. We might want to restrict click events to some `selection.distance`.